gcs-action/src/main.ts

```diff
@@ -0,0 +1,61 @@
+import * as core from '@actions/core'
+import * as gcs from './gcs'
+async function run(): Promise<void> {
+  try {
+    const folderToUpload = core.getInput('folder-to-upload')
+    const fileToUpload = core.getInput('file-to-upload')
+    const bucketName = core.getInput('bucket-name')
+    const objectKeyPrefix = core.getInput('object-key-prefix')
+    const clearExistingFilesFirst: boolean =
+      core.getInput('clear-existing-files-first').toLowerCase() !== 'false'
+
+    if (folderToUpload && !fileToUpload) {
```

Comment:
This `if` and the corresponding branching logic could be better split into other functions. That could help with testability. That being said, I don't think the user should be forced to decide if they are uploading a file vs a folder. I think we can detect this in the action based on the input path.

Reply:
+1 to the above. There's a lot of moving parts embedded in this run() method which can be split out into functions and tested orthogonally with unit tests.

Reply:
I have split it out, so that we can test most of it with unit tests. PTAL